### PR TITLE
AppCredentials construction: Enable extension of credential construction

### PR DESCRIPTION
**Pull request by carlosscastro**
AppCredentials construction: Enable extension of credential construction by sublcassing of adapter and http client. Side effect: promoting 4 private members to protected.

Instead of building new functionality for custom credentials in the adapter (and adding more constructors to the problem), we allow subclasses to customize how credentials are created. This enables scenarios like multitenant bots with custom creds, etc.

All the lines changed only boil down to 1) refactoring the creation of the AppCredentials to a protected virtual method, 2) promoting 4 private members to protected for subclasses to be able to access them. We make them properties since our policy does not allow protected members :)

### CODE REVIEW

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.ApplicationInsights.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.ApplicationInsights.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.AspNet.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.AspNet.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Testing.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Testing/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Streaming.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Streaming/4.6.3)

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 1]
```diff
@@ -1000,6 +1017,19 @@ protected virtual Task<ResourceResponse> ProcessOutgoingActivityAsync(ITurnConte
             throw new NotImplementedException();
         }
 
+        /// <summary>
+        /// Logic to build an <see cref="AppCreentials"/> object to be used to acquire tokens
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Spelling of Creentials

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 2]
```diff
@@ -1000,6 +1017,19 @@ protected virtual Task<ResourceResponse> ProcessOutgoingActivityAsync(ITurnConte
...
+        /// for this HttpClient.
+        /// </summary>
+        /// <param name="appId">The application id.</param>
+        /// <param name="oAuthScope">The optional OAuth scope.</param>
+        /// <returns>The app credentials to be used to acquire tokens.</returns>
+        protected virtual async Task<AppCredentials> BuildCredentialsAsync(string appId, string oAuthScope = null)
+        {
+            var appPassword = await CredentialProvider.GetAppPasswordAsync(appId).ConfigureAwait(false);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Should this throw if the appId is Null/Empty?

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 3]
```diff
@@ -1000,6 +1017,19 @@ protected virtual Task<ResourceResponse> ProcessOutgoingActivityAsync(ITurnConte
...
+            return ChannelProvider != null && ChannelProvider.IsGovernment() ? new MicrosoftGovernmentAppCredentials(appId, appPassword, HttpClient, Logger) : new MicrosoftAppCredentials(appId, appPassword, HttpClient, Logger, oAuthScope);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: This line confuses me, as I can't figure out what it's doing. 

Can you break it down a bit and add some comments. For example, why does the OAuthScope only get passed into the App, but not Government, Credentials?

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotFrameworkHttpClient.cs [hunk 4]
```diff
@@ -104,6 +109,19 @@ public async Task<InvokeResponse> PostActivityAsync(string fromBotId, string toB
...
         }
 
+        /// <summary>
+        /// Logic to build an <see cref="AppCreentials"/> object to be used to acquire tokens
+        /// for this HttpClient.
+        /// </summary>
+        /// <param name="appId">The application id.</param>
+        /// <param name="oAuthScope">The optional OAuth scope.</param>
+        /// <returns>The app credentials to be used to acquire tokens.</returns>
+        protected virtual async Task<AppCredentials> BuildCredentialsAsync(string appId, string oAuthScope = null)
```
Comment (cleemullins) [inline libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotFrameworkHttpClient.cs]: This seems the same as the code in the BotFrameworkAdapter above. Why do we need it in 2 places?

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 2 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: No, app id can be null or empty (it actually is in multiple test scnearios). Thanks for checking

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 3 above)
Comment (carlosscastro) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Yes, this is a legacy line. I'll add comments for now but I don't like the way GovernmentCreds inherits from the password specific ones, we have some improvement plans for R8 here.

File: libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotFrameworkHttpClient.cs (hunk 4 above)
Comment (carlosscastro) [reply libraries/integration/Microsoft.Bot.Builder.Integration.AspNet.Core/BotFrameworkHttpClient.cs]: The other code is in the new HttpClient which might has no coupling to the adapter at this point, and had a copy of this code. We will unify the code but need some more planning on what the right place is. Might do in a separate, R8 PR so I get exposure of this code from customers.

**Comment by fuselabs** [comment]
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.Luis.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.Luis/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.AI.QnA.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.AI.QnA/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.ApplicationInsights.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.ApplicationInsights/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Azure.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Azure/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Dialogs.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Dialogs/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.ApplicationInsights.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.ApplicationInsights.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Integration.AspNet.Core.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Integration.AspNet.Core/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.TemplateManager.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.TemplateManager/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Builder.Testing.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Builder.Testing/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Configuration.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Configuration/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Connector.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Connector/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Schema.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Schema/4.6.3)
:heavy_check_mark: No Binary Compatibility issues for **Microsoft.Bot.Streaming.dll** compared against [version 4.6.3](https://www.nuget.org/packages/Microsoft.Bot.Streaming/4.6.3)

**Pull request merged**